### PR TITLE
Support to bootstrap from a user-supplied security.json and minor refactorings to pass auth headers through a Context

### LINKED ISSUE #355: Bootstrap security.json from user-supplied ConfigMap
Off-shoot from #331 

Split out the non-OIDC related changes, such as supporting a user-supplied ConfigMap work in #346

### PR DESCRIPTION
**Pull request by thelabdude**
Resolves #355 

Includes the non-OIDC related security changes from #346

### CODE REVIEW

File: controllers/util/solr_security_util.go [hunk 1]
```diff
@@ -43,101 +44,134 @@ const (
...
+		err = reader.Get(ctx, types.NamespacedName{Name: instance.SecurityBootstrapSecretName(), Namespace: instance.Namespace}, bootstrapSecret)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				return nil, err
+			} // else perhaps the user deleted it after security was bootstrapped ... this is ok but may trigger a restart on the STS
+		} else {
+			// stash this so we can configure the setup-zk initContainer to bootstrap the security.json in ZK
 			security.SecurityJson = string(bootstrapSecret.Data[SecurityJsonFile])
-			basicAuthSecret = authSecret
+			security.SecurityJsonSrc = &corev1.EnvVarSource{
```
Comment (HoustonPutman) [inline controllers/util/solr_security_util.go:137]: Nice. love it.

File: controllers/util/solr_security_util.go [hunk 2]
```diff
@@ -43,101 +44,134 @@ const (
...
+		return nil, err
+	}
+
+	err := ValidateBasicAuthSecret(basicAuthSecret)
+	if err != nil {
+		return nil, err
+	}
+	security.CredentialsSecret = basicAuthSecret
+
+	// is there a user-provided security.json in a ConfigMap?
```
Comment (HoustonPutman) [inline controllers/util/solr_security_util.go:165]:
```suggestion
	// is there a user-provided security.json in a Secret?
```

File: controllers/util/solr_security_util.go [hunk 3]
```diff
@@ -43,101 +44,134 @@ const (
...
-				} // else perhaps the user deleted it after security was bootstrapped ... this is ok but may trigger a restart on the STS
-			} else {
-				// stash this so we can configure the setup-zk initContainer to bootstrap the security.json in ZK
-				security.SecurityJson = string(bootstrapSecret.Data[SecurityJsonFile])
-			}
-		}
-	}
+		security.SecurityJson = securityJson
+		security.SecurityJsonSrc = &corev1.EnvVarSource{SecretKeyRef: sec.BootstrapSecurityJson}
+	} // else no user-provided configMap, no sweat for us
```
Comment (HoustonPutman) [inline controllers/util/solr_security_util.go:174]:
```suggestion
	} // else no user-provided secret, no sweat for us
```

File: controllers/util/solr_security_util.go [hunk 4]
```diff
@@ -431,3 +483,22 @@ func useSecureProbe(solrCloud *solr.SolrCloud, probe *corev1.Probe, mountPath st
 		probe.TimeoutSeconds = 5
 	}
 }
+
+// Called during reconcile to load the security.json from a user-supplied ConfigMap
```
Comment (HoustonPutman) [inline controllers/util/solr_security_util.go:487]:
```suggestion
// Called during reconcile to load the security.json from a user-supplied Secret
```

**Review by HoustonPutman — APPROVED** [review]
Ok this looks good!

Was stumped a bit [here](https://github.com/apache/solr-operator/pull/356/files#diff-43592bd9457c52017b2b78f791395d14081935874a852a1d99a2e48d867c3e14R126) but I think it looks good (and I think I understand it haha).

Only things left to do are:

- [ ] Changelog entry (in solr-operator Chart.yaml)
- [ ] Add the option to the *solr* helm chart & helm chart documentation.

File: controllers/solrcloud_controller_basic_auth_test.go [hunk 5]
```diff
@@ -144,6 +173,34 @@ var _ = FDescribe("SolrCloud controller - Basic Auth", func() {
 			expectStatefulSetBasicAuthConfig(ctx, solrCloud, false)
 		})
 	})
+
+	FContext("User Provided Credentials and security.json ConfigMap", func() {
```
[the remark this answers is not in the document]
Comment (thelabdude) [reply controllers/solrcloud_controller_basic_auth_test.go]: ugh, sorry @HoustonPutman ... moved a little too fast with that conversion! thanks for catching those I missed.

**Pull request merged**